---
title: "scRNAseq analysis of Levesque’s data"
author: "Jonathan Haab"
date: "22 12 2020"
output: html_document
---
# from pipeline example described by Pierre-Luc Germain, course sta426 UZH

# Loading necessary libraries

```{r}
suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(scran)
  library(scater)
  library(batchelor)
  library(scDblFinder)
  library(sctransform)
  library(muscat)
  library(SEtools)
  library(cowplot)
  library(BiocParallel)
  library(ComplexHeatmap)
  
  library(DropletUtils) # for read10xCounts
  library(Matrix) # to handle sparse matrix
})
```

# Preprocessing & Clustering

## Loading the data

```{r data-loading}
## data can be downloaded from : http://imlspenticton.uzh.ch/dump/files_for_levesque.tar

# set local path
local.path <- getwd()
setwd(local.path)
patient1_HS.path <- file.path("data", "patient1_HS")
patient1_SCC.path <- file.path("data", "patient1_SCC")

# read in filtered data
fnameHS <- file.path(patient1_HS.path, "outs/filtered_feature_bc_matrix")
fnameSCC <- file.path(patient1_SCC.path, "outs/filtered_feature_bc_matrix")

# single cell experiment
sce.patient1_HS <- read10xCounts(fnameHS, col.names=TRUE)
sce.patient1_SCC <- read10xCounts(fnameSCC, col.names=TRUE)

sce.patient1_HS
sce.patient1_SCC
```

```{r split-data}
# only select the gene expression data, drop the antibody capture
#sce.patient1_HS <- sce.patient1_HS[rowData(sce.patient1_HS)$Type == "Gene Expression"]
#sce.patient1_SCC <- sce.patient1_SCC[rowData(sce.patient1_SCC)$Type == "Gene Expression"]

# Split the data, store ADT in alternative experiment
sce.patient1_HS <- splitAltExps(sce.patient1_HS, rowData(sce.patient1_HS)$Type)
sce.patient1_SCC <- splitAltExps(sce.patient1_SCC, rowData(sce.patient1_SCC)$Type)

# Coerce sparse matrix for ADT into a dense matrix
counts(altExp(sce.patient1_HS)) <- as.matrix(counts(altExp(sce.patient1_HS)))
counts(altExp(sce.patient1_SCC)) <- as.matrix(counts(altExp(sce.patient1_SCC)))

```


## Normalization & reduction
Note : The presence of composition biases already implies strong differences in expression profiles, so changing the normalization strategy is unlikely to affect the outcome of a clustering procedure.

Library size normalization : assumes that there is no “imbalance” in the differentially expressed (DE) genes between any pair of cells. However, balanced DE is not generally present in scRNA-seq applications, which means that library size normalization may not yield accurate normalized expression values for downstream analyses. In practice, normalization accuracy is not a major consideration for exploratory scRNA-seq data analyses. Composition biases do not usually affect the separation of clusters, only the magnitude - and to a lesser extent, direction - of the log-fold changes between clusters or cell types. As such, library size normalization is usually sufficient in many applications where the aim is to identify clusters and the top markers that define each cluster.

```{r library_size_normalization}
#exlude genes that appeared in less than 5 cells
#sce.patient1_HS <- sce[rowData(sce.patient1_HS)$detected >= 5,]

keep_feature <- rowSums(counts(sce.patient1_HS) > 0) > 4
sce.patient1_HS <- sce.patient1_HS[keep_feature,]

lib.sf.HS <- librarySizeFactors(sce.patient1_HS)
summary(lib.sf.HS)
hist(log10(lib.sf.HS), xlab="Log10[Size factor]", col='grey80')
```


```{r scaling_logtransform}
set.seed(100)
clust.patient1_HS <- quickCluster(sce.patient1_HS) 
sce.patient1_HS <- computeSumFactors(sce.patient1_HS, cluster=clust.patient1_HS, min.mean=0.1)
# downsample to correct the effect of library size
sce.patient1_HS <- logNormCounts(sce.patient1_HS) 
#sce.patient1_HS <- logNormCounts(sce.patient1_HS, downsample=TRUE) 
assayNames(sce.patient1_HS)

sce.patient1_HS <- runPCA(sce.patient1_HS)
plotPCA(sce.patient1_HS)
```
=> log transfo doesn't give satisfying results, we'll try variance-stabilizing transformation

```{r vst}
vst.patient1_HS <- suppressWarnings(sctransform::vst(counts(sce.patient1_HS)))
logcounts(sce.patient1_HS) <- as.matrix(logcounts(sce.patient1_HS))
nrow(vst.patient1_HS$y)
ncol(vst.patient1_HS$y)
nrow(counts(sce.patient1_HS))
logcounts(sce.patient1_HS, withDimnames=FALSE) <- vst.patient1_HS$y # ERROR
#assay(sce.patient1_HS, "vst") <- vst.patient1_HS$y
```

# ignore what's below

```{r}
var(logcounts(sce.patient1_HS))
# get highly-variable genes
hvgHS <- row.names(sce)[order(vst$gene_attr$residual_variance, decreasing=TRUE)[1:2000]]
hvgSCC <-
```

## Clustering
```{r}
gHS <- buildKNNGraph(sce.patient1_HS, BNPARAM=AnnoyParam(), use.dimred="PCA", k=30)
sce.patient1_HS$cluster <- as.factor(cluster_louvain(gHS)$membership)


plot_grid( plotTSNE(sce.patient1_HS, colour_by="cluster", text_by="cluster"),
           plotUMAP(sce.patient1_HS, colour_by="cluster", text_by="cluster") )
```
